# K-nearest Neighbors

In [1]:
import pandas as pd
home_listings = pd.read_csv("listings.csv")
home_listings.head()

id                           listing_url       scrape_id last_scraped  \
0  7087327  https://www.airbnb.com/rooms/7087327  20151002231825   2015-10-03   
1   975833   https://www.airbnb.com/rooms/975833  20151002231825   2015-10-03   
2  8249488  https://www.airbnb.com/rooms/8249488  20151002231825   2015-10-03   
3  8409022  https://www.airbnb.com/rooms/8409022  20151002231825   2015-10-03   
4  8411173  https://www.airbnb.com/rooms/8411173  20151002231825   2015-10-03   

                                 name  \
0  Historic DC Condo-Walk to Capitol!   
1     Spacious Capitol Hill Townhouse   
2    Spacious/private room for single   
3    A wonderful bedroom with library   
4              Downtown Silver Spring   

                                             summary  \
0  Professional pictures coming soon! Welcome to ...   
1                                                NaN   
2  This is an ideal room for a single traveler th...   
3  Prime location right on the Potomac River in W...   
4  Hi travellers! I live in this peaceful spot, b...   

                                               space  \
0                                                NaN   
1  Beautifully renovated Capitol Hill townhouse. ...   
2                                                NaN   
3                                                NaN   
4  This is a 750 sq ft 1 bedroom 1 bathroom.  Whi...   

                                         description experiences_offered  \
0  Professional pictures coming soon! Welcome to ...                none   
1  Beautifully renovated Capitol Hill townhouse. ...                none   
2  This is an ideal room for a single traveler th...                none   
3  Prime location right on the Potomac River in W...                none   
4  Hi travellers! I live in this peaceful spot, b...                none   

                               neighborhood_overview        ...         \
0                                                NaN        ...          
1                                                NaN        ...          
2                                                NaN        ...          
3                                                NaN        ...          
4  Silver Spring is booming.  You can walk to a n...        ...          

  review_scores_value requires_license license  \
0                 NaN                f     NaN   
1                 9.0                f     NaN   
2                 NaN                f     NaN   
3                 NaN                f     NaN   
4                 NaN                f     NaN   

                 jurisdiction_names instant_bookable cancellation_policy  \
0  DISTRICT OF COLUMBIA, WASHINGTON                f            flexible   
1  DISTRICT OF COLUMBIA, WASHINGTON                f              strict   
2                               NaN                f            flexible   
3  DISTRICT OF COLUMBIA, WASHINGTON                f            flexible   
4                               NaN                f            flexible   

   require_guest_profile_picture require_guest_phone_verification  \
0                              f                                f   
1                              f                                f   
2                              f                                f   
3                              f                                f   
4                              f                                f   

  calculated_host_listings_count reviews_per_month  
0                             18               NaN  
1                              1              2.11  
2                              1              1.00  
3                              1               NaN  
4                              1               NaN  

[5 rows x 92 columns]

### Similarity metric 
Using Euclidean distance and a univariate case(comparing single feature i.e. no of rooms)

In [2]:
import numpy as np
home_listings['distance'] = abs(home_listings['accommodates']-3)
home_listings['distance'].value_counts()

1     2294
2      503
0      461
3      279
5       73
4       35
7       22
6       17
9       12
13       8
8        7
12       6
11       4
10       2
Name: distance, dtype: int64

If we are looking for places having 3 rooms, there are 461 results(spaces) having 3 rooms. If we want to take the vaue 
of K say 5, we just can't sort and choose as the result will be biased. So we use random.

In [3]:
home_listings = home_listings.loc[np.random.permutation(len(home_listings))]
home_listings = home_listings.sort_values('distance')
print(home_listings.iloc[0:10]['price'])

380     $219.00
1486    $110.00
3341    $120.00
3060    $165.00
2919    $150.00
2312    $100.00
3572    $149.00
3100     $95.00
3226     $91.00
3169    $139.00
Name: price, dtype: object


### Cleaning the price column for computation

In [4]:
home_listings['price'] = home_listings['price'].str.replace(",","")
home_listings['price'] = home_listings['price'].str.replace("$","")
home_listings['price'] = home_listings['price'].astype('float')


In [6]:
mean_price = np.mean(home_listings['price'][0:5])
mean_price

152.8

### Generalizing

In [8]:

def predict_price(new_listing):
    temp_df = home_listings.copy()
    temp_df['distance'] = np.abs(temp_df['accommodates']-new_listing)
    temp_df = temp_df.sort_values('distance')
    first_five = temp_df['price'].iloc[0:5]
    mean_val = first_five.mean()
    return(mean_val)

acc_one = predict_price(1)
acc_two = predict_price(2)
acc_four = predict_price(4)

print(acc_one)
print(acc_two)
print(acc_four)

75.2
88.8
167.4
